src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs

```diff
@@ -891,21 +809,143 @@ EnumeratorResult createPatternBasedEnumeratorResult(ref ForEachEnumeratorInfo.Bu
                 return EnumeratorResult.FailedAndReported;
             }
+        }
+
+        private EnumeratorResult SatisfiesIEnumerableInterfaces(ref ForEachEnumeratorInfo.Builder builder, ref BoundExpression collectionExpr, bool isAsync, DiagnosticBag diagnostics, BoundExpression unwrappedCollectionExpr, TypeSymbol unwrappedCollectionExprType)
+        {
+            if (!AllInterfacesContainsIEnumerable(ref builder, unwrappedCollectionExprType, isAsync, diagnostics, out bool foundMultipleGenericIEnumerableInterfaces))
+            {
+                return EnumeratorResult.FailedNotReported;
+            }
+
+            collectionExpr = unwrappedCollectionExpr;
+            if (ReportConstantNullCollectionExpr(collectionExpr, diagnostics))
+            {
+                return EnumeratorResult.FailedAndReported;
+            }
+
+            CSharpSyntaxNode errorLocationSyntax = _syntax.Expression;
+
+            if (foundMultipleGenericIEnumerableInterfaces)
+            {
+                diagnostics.Add(isAsync ? ErrorCode.ERR_MultipleIAsyncEnumOfT : ErrorCode.ERR_MultipleIEnumOfT, errorLocationSyntax.Location, unwrappedCollectionExprType,
+                    isAsync ?
+                        this.Compilation.GetWellKnownType(WellKnownType.System_Collections_Generic_IAsyncEnumerable_T) :
+                        this.Compilation.GetSpecialType(SpecialType.System_Collections_Generic_IEnumerable_T));
+                return EnumeratorResult.FailedAndReported;
+            }
+
+            Debug.Assert((object)builder.CollectionType != null);
 
-            bool reportConstantNullCollectionExpr(BoundExpression collectionExpr)
+            NamedTypeSymbol collectionType = (NamedTypeSymbol)builder.CollectionType;
+            if (collectionType.IsGenericType)
             {
-                if (collectionExpr.ConstantValue is { IsNull: true })
+                // If the type is generic, we have to search for the methods
+                builder.ElementTypeWithAnnotations = collectionType.TypeArgumentsWithAnnotationsNoUseSiteDiagnostics.Single();
+
+                MethodSymbol getEnumeratorMethod;
+                if (isAsync)
                 {
-                    // Spec seems to refer to null literals, but Dev10 reports anything known to be null.
-                    diagnostics.Add(ErrorCode.ERR_NullNotValid, _syntax.Expression.Location);
-                    return true;
+                    Debug.Assert(IsIAsyncEnumerable(collectionType.OriginalDefinition));
+
+                    getEnumeratorMethod = (MethodSymbol)GetWellKnownTypeMember(Compilation, WellKnownMember.System_Collections_Generic_IAsyncEnumerable_T__GetAsyncEnumerator,
+                        diagnostics, errorLocationSyntax.Location, isOptional: false);
+
+                    // Well-known members are matched by signature: we shouldn't find it if it doesn't have exactly 1 parameter.
+                    Debug.Assert(getEnumeratorMethod.ParameterCount == 1);
```

Comment:
>getEnumeratorMethod [](start = 33, length = 19)

Possibly dereferencing a null ref? Is there a test hole?

Reply:
Looks like we're missing the standard missing member tests for all of the async tests. I'll add them.

---
In reply to: [550225429](https://github.com/dotnet/roslyn/pull/50105#discussion_r550225429) [](ancestors = 550225429)